klippy/extras/tmc2240.py

```diff
@@ -0,0 +1,399 @@
+    "w1":                       0x03 << 2,
+    "w0":                       0x03 << 0,
+}
+Fields["MSLUTSTART"] = {
+    "start_sin":                0xFF << 0,
+    "start_sin90":              0xFF << 16,
+    "offset_sin90":             0xFF << 24,
+}
+Fields["MSCNT"] = {
+    "mscnt":                    0x3ff << 0
+}
+Fields["MSCURACT"] = {
+    "cur_a":                    0x1ff << 0,
+    "cur_b":                    0x1ff << 16
+}
+Fields["PWM_AUTO"] = {
+    "pwm_ofs_auto":             0xff << 0,
+    "pwm_grad_auto":            0xff << 16
+}
+Fields["PWMCONF"] = {
+    "pwm_ofs":                  0xFF << 0,
+    "pwm_grad":                 0xFF << 8,
+    "pwm_freq":                 0x03 << 16,
+    "pwm_autoscale":            0x01 << 18,
+    "pwm_autograd":             0x01 << 19,
+    "freewheel":                0x03 << 20,
+    "pwm_meas_sd_enable":       0x01 << 22,
+    "pwm_dis_reg_stst":         0x01 << 23,
+    "pwm_reg":                  0x0F << 24,
+    "pwm_lim":                  0x0F << 28
+}
+Fields["PWM_SCALE"] = {
+    "pwm_scale_sum":            0x3ff << 0,
+    "pwm_scale_auto":           0x1ff << 16
+}
+Fields["TPOWERDOWN"] = {
+    "tpowerdown":               0xff << 0
+}
+Fields["TPWMTHRS"] = {
+    "tpwmthrs":                 0xfffff << 0
+}
+Fields["TCOOLTHRS"] = {
+    "tcoolthrs":                0xfffff << 0
+}
+Fields["TSTEP"] = {
+    "tstep":                    0xfffff << 0
+}
+Fields["THIGH"] = {
+    "thigh":                    0xfffff << 0
+}
+Fields["DRV_CONF"] = {
+    "current_range":            0x03 << 0,
+    "slope_control":            0x03 << 4
+}
+Fields["ADC_VSUPPLY_AIN"] = {
+    "adc_vsupply":              0x1fff << 0,
+    "adc_ain":                  0x1fff << 16
+}
+Fields["ADC_TEMP"] = {
+    "adc_temp":                 0x1fff << 0
+}
+Fields["OTW_OV_VTH"] = {
+    "overvoltage_vth":          0x1fff << 0,
+    "overtempprewarning_vth":   0x1fff << 16
+}
+Fields["SG4_THRS"] = {
+    "sg4_thrs":                 0xFF << 0,
+    "sg4_filt_en":              0x01 << 8,
+    "sg4_angle_offset":         0x01 << 9
+}
+Fields["SG4_RESULT"] = {
+    "sg4_result":               0x3FF << 0
+}
+Fields["SG4_IND"] = {
+    "sg4_ind_0":                0xFF << 0,
+    "sg4_ind_1":                0xFF << 8,
+    "sg4_ind_2":                0xFF << 16,
+    "sg4_ind_3":                0xFF << 24
+}
+
+
+SignedFields = ["cur_a", "cur_b", "sgt", "pwm_scale_auto", "offset_sin90"]
+
+FieldFormatters = dict(tmc2130.FieldFormatters)
+FieldFormatters.update({
+    "s2vsa":            (lambda v: "1(ShortToSupply_A!)" if v else ""),
+    "s2vsb":            (lambda v: "1(ShortToSupply_B!)" if v else ""),
+    "adc_temp":         (lambda v: "0x%04x(%.1fC)" % (v, ((v - 2038) / 7.7))),
+})
+
+
+######################################################################
+# TMC stepper current config helper
+######################################################################
+
+MAX_CURRENT = 3.000
```

Comment:
I don't think this is correct. This is peak max current, while klipper uses RMS max current in all other places. The datasheet says the max RMS current is 2.1A, which is true when using Rref=12K and current_range=3, but a higher Rref will reduce the possible max current. So I think this value should also be calculated dynamically.